dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/ReferenceBeanManager.java

```diff
@@ -0,0 +1,253 @@
+     * @throws Exception
+     */
+    public void prepareReferenceBeans() throws Exception {
+        initialized = true;
+        for (ReferenceBean referenceBean : getReferences()) {
+            initReferenceBean(referenceBean, false);
+        }
+    }
+
+    /**
+     * NOTE: This method should only call after all dubbo config beans and all property resolvers is loaded.
+     *
+     * @param referenceBean
+     * @throws Exception
+     */
+    private void initReferenceBean(ReferenceBean referenceBean, boolean canCreateProxy) throws Exception {
+
+        if (referenceBean.getReferenceConfig() != null) {
+            return;
+        }
+
+        Environment environment = applicationContext.getEnvironment();
+        Map<String, Object> referenceProps = referenceBean.getReferenceProps();
+        if (referenceProps == null) {
+            MutablePropertyValues propertyValues = referenceBean.getPropertyValues();
+            if (propertyValues == null) {
+                throw new RuntimeException("ReferenceBean is invalid, missing 'propertyValues'");
+            }
+            referenceProps = toReferenceProps(propertyValues, environment);
+        }
+
+        //resolve placeholders
+        resolvePlaceholders(referenceProps, environment);
+
+        //create real ReferenceConfig
+        ReferenceConfig referenceConfig = ReferenceBeanBuilder.create(new AnnotationAttributes(new LinkedHashMap<>(referenceProps)), applicationContext)
+                .defaultInterfaceClass(referenceBean.getObjectType())
+                .build();
+
+        referenceBean.setReferenceConfig(referenceConfig);
+
+        // register ReferenceConfig
+        DubboBootstrap.getInstance().reference(referenceConfig);
+
+        //TODO add after DubboBootstrap is started
+        if (canCreateProxy && referenceConfig.shouldInit()) {
+            ReferenceConfigCache.getCache().get(referenceConfig);
+        }
+    }
+
+    private void resolvePlaceholders(Map<String, Object> referenceProps, PropertyResolver propertyResolver) {
+        for (Map.Entry<String, Object> entry : referenceProps.entrySet()) {
+            Object value = entry.getValue();
+            if (value instanceof String) {
+                String valueToResovle = (String) value;
+                entry.setValue(propertyResolver.resolveRequiredPlaceholders(valueToResovle));
+            } else if (value instanceof String[]) {
+                String[] strings = (String[]) value;
+                for (int i = 0; i < strings.length; i++) {
+                    strings[i] = propertyResolver.resolveRequiredPlaceholders(strings[i]);
+                }
+                entry.setValue(strings);
+            }
+        }
+    }
+
+    private Map<String, Object> toReferenceProps(MutablePropertyValues propertyValues, PropertyResolver propertyResolver) {
+        Map<String, Object> referenceProps;
+        referenceProps = new LinkedHashMap<>();
```

Comment:
shall we make this map be concurrentable? or return a unmodifiable map

Reply:
`Map<String, Object> referenceProps` is temporary data, only used once.